### DISCUSSION TITLE
7.x.x - Behaviors

### DISCUSSION THREAD

**Discussion by swingingtom** [felixmariotto/three-mesh-ui#208] [General]
In the 7.x.x branch, the concept of `Behaviors` has been launched. But its implementation and API are still unstable. This page is all about making them fast, easy to use, reliable and useful, together.

## Behavior idea
The idea behind `Behaviors` is that three-mesh-ui could provide some (a lot?) of easy to use 'Behavior' to users.     

### Technical concerns
If any of those are implemented in core, medium-term maintenance would suffer.

If any of those are implemented in core, all users (even without using any behaviors) will have to load more as behaviors will be implemented in core.

### Goal
Providing easy to use behavior, without making the core grow exponentially.
Allowing users to build their own behaviors to fit their project.

## Context
The perfect example is the `BestFit`, which is a nice feature implemented since 6.x. On the 6.x it is implemented in core, and required core to be adapted to that feature. 

### 6.x.x Implementation
`core/MeshUIComponent` : [L534](https://github.com/felixmariotto/three-mesh-ui/blob/1505a0ca1e0f83d37f49cc5131d46a378a14e7e7/src/components/core/MeshUIComponent.js#L534) + [L545](https://github.com/felixmariotto/three-mesh-ui/blob/1505a0ca1e0f83d37f49cc5131d46a378a14e7e7/src/components/core/MeshUIComponent.js#L545) + [L552](https://github.com/felixmariotto/three-mesh-ui/blob/1505a0ca1e0f83d37f49cc5131d46a378a14e7e7/src/components/core/MeshUIComponent.js#L552) + [L534](https://github.com/felixmariotto/three-mesh-ui/blob/1505a0ca1e0f83d37f49cc5131d46a378a14e7e7/src/components/core/MeshUIComponent.js#L609) + [L340](https://github.com/felixmariotto/three-mesh-ui/blob/1505a0ca1e0f83d37f49cc5131d46a378a14e7e7/src/components/core/MeshUIComponent.js#L340)    
`core/Block` : [L66](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/Block.js#L66)   
`core/InlineManager` : [L104](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/core/InlineManager.js#L104) + [L122](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/core/InlineManager.js#L122) + [L175](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/core/InlineManager.js#L175) + [L225](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/core/InlineManager.js#L225) + [L388](https://github.com/felixmariotto/three-mesh-ui/blob/4f28775c8852ee6b3872f333371c7a04b118ede9/src/components/core/InlineManager.js#L388)

### 6.x.x API
```javascript
// in 6.x.x, API for bestFit is the same as any three-mesh-ui property: .set()
const block = new ThreeMeshUI.Block({bestFit:'grow'});
```

If others useful `Behavior` require the same, the core would literally implodes. This is why on 7.x.x `BestFit` feature is has been refactored as a Behavior.

### 7.x.x Implementation
_**No core components has knowledge that bestFit may append**_
`examples/behaviors/size/BestFitBehavior` : The behavior holds any of code required for bestFit [source](https://github.com/felixmariotto/three-mesh-ui/blob/7.x.x/examples/behaviors/size/BestFitBehavior.js)

BestFitBehavior will add new methods to its subject: https://github.com/felixmariotto/three-mesh-ui/blob/1c5cb6a1f8b4d94bcbf2460e156809c8f1a46dd3/examples/behaviors/size/BestFitBehavior.js#L10
BestFitBehavior will override its subject `parseParams()` method: https://github.com/felixmariotto/three-mesh-ui/blob/1c5cb6a1f8b4d94bcbf2460e156809c8f1a46dd3/examples/behaviors/size/BestFitBehavior.js#L40-L44

### 7.x.x API
```javascript
const innerBlock = new ThreeMeshUI.Block( { width: 0.5, height: 0.3 } );

// 7.x.x : Add a behavior on Block that need bestFit
const innerBestFit = new BestFitBehavior( innerContainer, 'grow' );

// start the behavior
innerBestFit.attach();

// stop the behavior
innerBestFit.detach();

// just execute the behavior on one frame
innerBestFit.act(); //* currently not allowed on bestFit behavior
```

### Existing Behaviors on 7.x.x
- **BestFit:** Automatically resize font to fit : https://felixmariotto.github.io/three-mesh-ui/7.x.x/#ex__best_fit
- **BoundsUV:** Apply UV from another object size : https://felixmariotto.github.io/three-mesh-ui/7.x.x/#ex__bounds_uv
- **TextLayout:** (Utils) Underlay typographic values : lineHeight, baseLine, medianLine on Text : 
![image](https://user-images.githubusercontent.com/17314210/176849931-34bc67e3-d1cf-4c17-a9f1-b50613f62da0.png)
- **BoxLayout:** (Utils) Overlay box layout values : margin, border, padding such devTool : 
![image](https://user-images.githubusercontent.com/17314210/176852458-cc844d58-3336-401f-9682-e7a2ccd64c69.png)
- **BoxAnchors:** (Utils) Overlay box corners

### Foreseen Behaviors
- **OverflowPath:** Automatically split textContent through multiple components : 
![image](https://user-images.githubusercontent.com/17314210/176854566-ed58f1c6-015c-4345-9646-ae5285d6604a.png)
- **AutoWidth:** Adjust a parent size according to its children (#207)

**Comment by swingingtom** [comment]
# API Discussion
Im currently no totally convince about the behavior API implementation.

The current implementation seems totally disconnected from components, and leave an orphan Behavior element.

## Alternatives
### From component
```javascript
myBlock.addBehavior( new Behavior(...) );
```
Where this could be useful as a `component.dispose()` would be able to automatically remove any existing behavior, it would challenge us when a Behavior requires multiComponents : BoundsUV, OverflowPath, ....

Any inputs?
